tests/FSharp.Compiler.ComponentTests/ErrorMessages/ExtendedDiagnosticDataTests.fs

```diff
@@ -0,0 +1,203 @@
+
+[<Fact>]
+let ``TypeMismatchDiagnosticExtendedData 02`` () =
+    FSharp """
+let x, y = 1
+"""
+    |> typecheckResults
+    |> checkDiagnostic
+       (1, "This expression was expected to have type\n    ''a * 'b'    \nbut here has type\n    'int'    ")
+       (fun (typeMismatch: TypeMismatchDiagnosticExtendedData) ->
+        let displayContext = typeMismatch.DisplayContext
+        Assert.Equal(DiagnosticContextInfo.NoContext, typeMismatch.ContextInfo)
+        Assert.Equal("obj * obj", typeMismatch.ExpectedType.Format(displayContext))
+        Assert.Equal("int", typeMismatch.ActualType.Format(displayContext)))
+
+[<Fact>]
+let ``TypeMismatchDiagnosticExtendedData 03`` () =
+    FSharp """
+if true then 5
+"""
+    |> typecheckResults
+    |> checkDiagnostic
+       (1, "This 'if' expression is missing an 'else' branch. Because 'if' is an expression, and not a statement, add an 'else' branch which also returns a value of type 'int'.")
+       (fun (typeMismatch: TypeMismatchDiagnosticExtendedData) ->
+        let displayContext = typeMismatch.DisplayContext
+        Assert.Equal(DiagnosticContextInfo.OmittedElseBranch, typeMismatch.ContextInfo)
+        Assert.Equal("unit", typeMismatch.ExpectedType.Format(displayContext))
+        Assert.Equal("int", typeMismatch.ActualType.Format(displayContext)))
+
+[<Fact>]
+let ``TypeMismatchDiagnosticExtendedData 04`` () =
+    FSharp """
+1 :> string
+"""
+    |> typecheckResults
+    |> checkDiagnostic
+       (193, "Type constraint mismatch. The type \n    'int'    \nis not compatible with type\n    'string'    \n")
+       (fun (typeMismatch: TypeMismatchDiagnosticExtendedData) ->
+        let displayContext = typeMismatch.DisplayContext
+        Assert.Equal(DiagnosticContextInfo.NoContext, typeMismatch.ContextInfo)
+        Assert.Equal("string", typeMismatch.ExpectedType.Format(displayContext))
+        Assert.Equal("int", typeMismatch.ActualType.Format(displayContext)))
+
+[<Fact>]
+let ``TypeMismatchDiagnosticExtendedData 05`` () =
+    FSharp """
+match 0 with
+| 0 -> 1
+| 1 -> "a"
+"""
+    |> typecheckResults
+    |> checkDiagnostic
+       (1, "All branches of a pattern match expression must return values implicitly convertible to the type of the first branch, which here is 'int'. This branch returns a value of type 'string'.")
+       (fun (typeMismatch: TypeMismatchDiagnosticExtendedData) ->
+        let displayContext = typeMismatch.DisplayContext
+        Assert.Equal(DiagnosticContextInfo.FollowingPatternMatchClause, typeMismatch.ContextInfo)
+        Assert.Equal("int", typeMismatch.ExpectedType.Format(displayContext))
+        Assert.Equal("string", typeMismatch.ActualType.Format(displayContext)))
+
+//TODO: FollowingPatternMatchClause should be provided for type equation diagnostics come from a return type only
+[<Fact>]
+let ``TypeMismatchDiagnosticExtendedData 05 - TODO: fix wrong context`` () =
```

Comment:
Do you plan to address this in this PR or separately?

Reply:
Separately, I suppose.

Reply:
![i-trust-you-kaulder](https://github.com/dotnet/fsharp/assets/5451366/bfc07135-b386-434c-a457-694ef05e4946)